// ngx-fabric-wrapper/esm2020/lib/fabric.module.mjs
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FabricComponent } from './fabric.component';
import { FabricDirective } from './fabric.directive';
import * as i0 from "@angular/core";
export class FabricModule {
}
FabricModule.ɵfac = i0.ɵɵngDeclareFactory({ minVersion: "12.0.0", version: "15.2.0", ngImport: i0, type: FabricModule, deps: [], target: i0.ɵɵFactoryTarget.NgModule });
FabricModule.ɵmod = i0.ɵɵngDeclareNgModule({ minVersion: "14.0.0", version: "15.2.0", ngImport: i0, type: FabricModule, declarations: [FabricComponent, FabricDirective], imports: [CommonModule], exports: [CommonModule, FabricComponent, FabricDirective] });
FabricModule.ɵinj = i0.ɵɵngDeclareInjector({ minVersion: "12.0.0", version: "15.2.0", ngImport: i0, type: FabricModule, imports: [CommonModule, CommonModule] });
i0.ɵɵngDeclareClassMetadata({ minVersion: "12.0.0", version: "15.2.0", ngImport: i0, type: FabricModule, decorators: [{
            type: NgModule,
            args: [{
                    declarations: [FabricComponent, FabricDirective],
                    imports: [CommonModule],
                    exports: [CommonModule, FabricComponent, FabricDirective],
                    providers: []
                }]
        }] });
